File: src/Leitor/LeitorAvaliacao.java
package Leitor;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

import Modelo.Avaliacao;
import Modelo.Estudante;

public class LeitorAvaliacao extends ILeitor implements Serializable{
    
    private static final long serialVersionUID = 1L;
    private static LeitorAvaliacao leitor;

    private LeitorAvaliacao(){}

    public static LeitorAvaliacao obterInstancia(){
        if (leitor == null){
            leitor = new LeitorAvaliacao();
            return leitor;
        }
        else return leitor;
    }

    public ArrayList<Avaliacao> busca(List<Avaliacao> avaliacoes,Estudante estudante){
        ArrayList<Avaliacao> lista = new ArrayList<>();
        for(Avaliacao a : avaliacoes){
            if(a.obterAluno() == estudante){
                lista.add(a);
            }
        }
        return lista;
    }

    public Avaliacao ler(Scanner scan, Estudante aluno) throws Exception {
        
        float nota;

        try {
            nota = scan.nextFloat();
        } catch (Exception e){
            throw new Exception("Dado inválido: " + scan.next() + ".");
        }

        return new Avaliacao(aluno, nota);
    }
}

File: cpp_version/Leitor/LeitorAvaliacao.cpp
#include "LeitorAvaliacao.h"

LeitorAvaliacao* LeitorAvaliacao::leitor = nullptr;

LeitorAvaliacao::LeitorAvaliacao(){}

LeitorAvaliacao* LeitorAvaliacao::obterInstancia(){
    if (leitor == NULL){
        leitor = new LeitorAvaliacao();
        return leitor;
    }
    else return leitor;
}

vector<Avaliacao*> LeitorAvaliacao::busca(vector<Avaliacao*> avaliacoes, Estudante* estudante){
    vector<Avaliacao*> lista;
    for(Avaliacao* a : avaliacoes){
        if(a->obterAluno()->obterMatricula() == estudante->obterMatricula())
            lista.push_back(a);
    }
    return lista;
}

Avaliacao* LeitorAvaliacao::ler(vector<string> vec,Estudante aluno){
    double nota;
    string trash;

    try{
        nota = stof(vec[3]);
    } catch (...) {
        throw ExcecaoDado("Dado inválido: " + vec[3] + ".");
    }

    return new Avaliacao(aluno, nota);
}